File: quickstarts/go/main-eth.go
// CODE:BEGIN:quickstarts_go_ethereum_section1
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"

	pb "github.com/dfuse-io/docs/quickstarts/go/pb"

	"github.com/tidwall/gjson"
	"golang.org/x/oauth2"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/oauth"
)

func main() {

	dfuseAPIKey := os.Getenv("DFUSE_API_KEY")
	if dfuseAPIKey == "" || dfuseAPIKey == "your dfuse api key here" {
		panic("you must specify a DFUSE_API_KEY environment variable")
	}

	token, _, err := getToken(dfuseAPIKey)
	panicIfError(err)

	client := createClient("mainnet.eth.dfuse.io:443", token)

	streamEthereum(context.Background(), client)
}

func panicIfError(err error) {
	if err != nil {
		panic(err)
	}
}

// CODE:END:quickstarts_go_ethereum_section1

// CODE:BEGIN:quickstarts_go_ethereum_section2
func getToken(apiKey string) (token string, expiration time.Time, err error) {
	reqBody := bytes.NewBuffer([]byte(fmt.Sprintf(`{"api_key":"%s"}`, apiKey)))
	resp, err := http.Post("https://auth.dfuse.io/v1/auth/issue", "application/json", reqBody)
	if err != nil {
		err = fmt.Errorf("unable to obtain token: %s", err)
		return
	}

	if resp.StatusCode != 200 {
		err = fmt.Errorf("unable to obtain token, status not 200, got %d: %s", resp.StatusCode, reqBody.String())
		return
	}

	if body, err := ioutil.ReadAll(resp.Body); err == nil {
		token = gjson.GetBytes(body, "token").String()
		expiration = time.Unix(gjson.GetBytes(body, "expires_at").Int(), 0)
	}
	return
}

// CODE:END:quickstarts_go_ethereum_section2

// CODE:BEGIN:quickstarts_go_ethereum_section3
func createClient(endpoint string, token string) pb.GraphQLClient {

	credential := oauth.NewOauthAccess(&oauth2.Token{AccessToken: token, TokenType: "Bearer"})
	transportCreds := credentials.NewClientTLSFromCert(nil, "")
	conn, err := grpc.Dial(endpoint,
		grpc.WithPerRPCCredentials(credential),
		grpc.WithTransportCredentials(transportCreds),
	)
	panicIfError(err)

	return pb.NewGraphQLClient(conn)
}

// CODE:END:quickstarts_go_ethereum_section3

// CODE:BEGIN:quickstarts_go_ethereum_section4
const operationETH = `subscription {
  searchTransactions(indexName:CALLS, query:"-value:0 type:call", lowBlockNum: -1) {
    undo cursor
    node { hash matchingCalls { from to value(encoding:ETHER) } }
  }
}`

type ethereumDocument struct {
	SearchTransactions struct {
		Cursor string
		Undo   bool
		Node   struct {
			Hash          string
			MatchingCalls []struct {
				From  string
				To    string
				Value string
			}
		}
	}
}

// CODE:END:quickstarts_go_ethereum_section4

// CODE:BEGIN:quickstarts_go_ethereum_section5
func streamEthereum(ctx context.Context, client pb.GraphQLClient) {

	executor, err := client.Execute(ctx, &pb.Request{Query: operationETH})
	panicIfError(err)

	for {
		resp, err := executor.Recv()
		panicIfError(err)

		if len(resp.Errors) > 0 {
			for _, err := range resp.Errors {
				fmt.Printf("Request failed: %s\n", err)
			}

			/* We continue here, but you could take another decision here, like exiting the process */
			continue
		}

		document := &ethereumDocument{}
		err = json.Unmarshal([]byte(resp.Data), document)
		panicIfError(err)

		result := document.SearchTransactions
		reverted := ""
		if result.Undo {
			reverted = " REVERTED"
		}

		for _, call := range result.Node.MatchingCalls {
			fmt.Printf("Transfer %s -> %s [%s Ether]%s\n", call.From, call.To, call.Value, reverted)
		}
	}
}

// CODE:END:quickstarts_go_ethereum_section5

File: quickstarts/python/main-eth.py
# CODE:BEGIN:quickstarts_python_ethereum_section1
try:
    # python3
    from http.client import HTTPSConnection
except ImportError:
    # python2
    from httplib import HTTPSConnection

import grpc
import json
import os
import ssl
import sys

from graphql import graphql_pb2_grpc
from graphql.graphql_pb2 import Request
# CODE:END:quickstarts_python_ethereum_section1

# CODE:BEGIN:quickstarts_python_ethereum_section2
def get_token(api_key):
    connection = HTTPSConnection("auth.dfuse.io")
    connection.request('POST', '/v1/auth/issue', json.dumps({"api_key": api_key}), {'Content-type': 'application/json'})
    response = connection.getresponse()

    if response.status != 200:
        raise Exception(" Status: %s reason: %s" % (response.status, response.reason))

    token = json.loads(response.read().decode())['token']
    connection.close()

    return token
# CODE:END:quickstarts_python_ethereum_section2

# CODE:BEGIN:quickstarts_python_ethereum_section3
def create_client(token, endpoint):
    channel = grpc.secure_channel(endpoint,
        credentials = grpc.composite_channel_credentials(
            grpc.ssl_channel_credentials(),
            grpc.access_token_call_credentials(token)
    ))

    return graphql_pb2_grpc.GraphQLStub(channel)
# CODE:END:quickstarts_python_ethereum_section3

# CODE:BEGIN:quickstarts_python_ethereum_section4
OPERATION_ETH = """subscription {
  searchTransactions(indexName: CALLS, query: "-value:0 type:call", lowBlockNum: -1) {
    undo cursor
    node { hash matchingCalls { from to value(encoding:ETHER) } }
  }
}"""
# CODE:END:quickstarts_python_ethereum_section4

# CODE:BEGIN:quickstarts_python_ethereum_section5
def stream_ethereum(client):
    # The client can be re-used for all requests, cache it at the appropriate level
    stream = client.Execute(Request(query = OPERATION_ETH))

    for rawResult in stream:
        if rawResult.errors:
            print("An error occurred")
            print(rawResult.errors)
        else:
            result = json.loads(rawResult.data)
            for call in result['searchTransactions']['node']['matchingCalls']:
                undo = result['searchTransactions']['undo']
                print("Transfer %s -> %s [%s Ether]%s" % (call['from'], call['to'], call['value'], " REVERTED" if undo else ""))
# CODE:END:quickstarts_python_ethereum_section5

# CODE:BEGIN:quickstarts_python_ethereum_section6
dfuse_api_key = os.environ.get("DFUSE_API_KEY")
if dfuse_api_key == None or dfuse_api_key == 'your dfuse api key here':
    raise Exception('you must specify a DFUSE_API_KEY environment variable')

token = get_token(dfuse_api_key)
# The client can be re-used for all requests, cache it at the appropriate level

client = create_client(token, 'mainnet.eth.dfuse.io:443')
stream_ethereum(client)
# CODE:END:quickstarts_python_ethereum_section6